DDD-12.md

```diff
@@ -0,0 +1,326 @@
+
+```java
+import io.debezium.engine.ChangeEvent;
+import jakarta.enterprise.context.ApplicationScoped;  
+
+
+@ApplicationScoped  
+class OrderListener {
+  
+    @DebeziumListener("order")  
+    public void listener(ChangeEvent<String, String> event) {  
+        /// some logic to apply 
+    }  
+}
+```
+
+or in batch:
+
+```java
+import io.debezium.engine.ChangeEvent;
+import jakarta.enterprise.context.ApplicationScoped;  
+
+
+@ApplicationScoped  
+class OrderListener {
+  
+    @DebeziumBatchListener("order")  
+    public void listener(List<ChangeEvent<String, String>> events) {  
+        /// some logic to apply
+    }  
+}
+```
+
+even listen only a certain type of event (using for example the configuration `skipped.operations`)
+
+```java
+import io.debezium.engine.InsertEvent;
+import jakarta.enterprise.context.ApplicationScoped;  
+import io.debezium.engine.quarkus.Operation.INSERT;
+
+@ApplicationScoped  
+class OrderListener {
+  
+    @DebeziumBatchListener("order", INSERT)  
+    public void listener(List<InsertEvent<String, String>> events) {  
+        /// some logic to apply
+    }  
+}
+```
+
+### Custom Debezium Data Converter
+
+It should be possible to receive events mapped as data classes like:
+
+```java
+public record Order(long id, String name, int price) {}
+```
+
+```java
+import io.debezium.engine.InsertEvent;
+import jakarta.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped  
+class OrderListener {  
+  
+    @DebeziumListener("order")  
+    public void listener(InsertEvent<String, Order> event) {  
+        /// some logic to apply
+    }  
+}
+```
+
+using something similar for the [quarkus kafka library](https://quarkus.io/guides/kafka#jackson-serialization):
+
+```java
+package com.acme.order.jackson;
+
+import io.quarkus.debezium.client.serialization.ObjectMapperDeserializer;
+
+public class OrderDeserializer extends ObjectMapperDeserializer<Order> {
+    public OrderDeserializer() {
+        super(Order.class);
+    }
+}
+```
+
+```properties
+quarkus.debezium.deserializer=com.acme.order.jackson.OrderDeserializer
+```
+
+### Quarkus Debezium SchemaChange Listener
+Debezium automatically detects and captures schema changes in the source database, such as adding or removing columns, modifying data types, or altering primary keys. These changes are parsed from the database's DDL statements and used to update Debezium's internal schema history, ensuring that change events reflect the current table structure. The Quarkus extension can expose a listener to such kind of event:
+
+```java
+import io.debezium.engine.InsertEvent;
+import jakarta.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped  
+class SchemaChangeListener {  
+  
+    @DebeziumSchemaChangeListener()  
+    public void listener(SchemaChangeEvent event) {  
+        /// some logic to apply
+    }  
+}
+```
+
+### Quarkus Debezium Snapshot Listener
```

Comment:
Maybe here we should generically caputure the [Notifications](https://debezium.io/documentation/reference/3.1/configuration/notification.html) here. WDYT?

Reply:
Well, if you are talking about `Custom Notification` (since sink, log, and JMX are already tailored to specific use cases), it depends. Providing access to the `ChangeEvent` is certainly [powerful](https://github.com/debezium/debezium/blob/6a37d34f627f878280f692d51e498c1c64840605/debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java#L224), but on the other hand, the [NotificationChannel](https://github.com/debezium/debezium/blob/6a37d34f627f878280f692d51e498c1c64840605/debezium-core/src/main/java/io/debezium/pipeline/notification/channels/NotificationChannel.java#L22) doesn't seem powerful enough to support custom deserializers/converters.

Reply:
No, I'm talking more about something like https://github.com/debezium/debezium/blob/main/debezium-core/src/main/java/io/debezium/pipeline/signal/channels/process/InProcessSignalChannel.java but for the Notifications. 

So we should provide a `in-process` notification channel and then the user can get can just start listening to those notifications in a provided listener.

Reply:
If I understand correctly, what do you mean is to have an annotation that can access to  notifications and hook particular notifications (like snapshot)? like:

```java
@ApplicationScoped  
class NotificationChannel {  

    @DebeziumNotificationHandler(SnapshotNotification.class)  
    public void handler(Notification notification) {
        /// some logic to apply
    }
}
```

Could you please provide an example?

Reply:
For now I think just having a way to listen to all notifications will be fine.
```java
@ApplicationScoped  
class NotificationChannel {  

    public void handler(Notification notification) {
        /// some logic to apply
    }
}
```